trie/triedb/pathdb/nodebufferlist.go

```diff
@@ -110,31 +115,220 @@ func newNodeBufferList(
 		}
 	}
 	base := newMultiDifflayer(limit, size, common.Hash{}, nodes, layers)
-	ele := newMultiDifflayer(limit, 0, common.Hash{}, make(map[common.Hash]map[string]*trienode.Node), 0)
-	nf := &nodebufferlist{
+
+	var (
+		nf  *nodebufferlist
+		err error
+	)
+	if recovery {
+		nf, err = newNodeBufferListForRecovery(db, freezer, base, limit, wpBlocks, rsevMdNum, dlInMd)
+		if err != nil {
+			log.Error("Failed to new node buffer list for recovery", "error", err)
+			return nil, err
+		}
+	} else {
+		ele := newMultiDifflayer(limit, 0, common.Hash{}, make(map[common.Hash]map[string]*trienode.Node), 0)
+		nf = &nodebufferlist{
+			db:              db,
+			wpBlocks:        wpBlocks,
+			rsevMdNum:       rsevMdNum,
+			dlInMd:          dlInMd,
+			limit:           limit,
+			base:            base,
+			head:            ele,
+			tail:            ele,
+			count:           1,
+			persistID:       rawdb.ReadPersistentStateID(db),
+			stopCh:          make(chan struct{}),
+			waitStopCh:      make(chan struct{}),
+			forceKeepCh:     make(chan struct{}),
+			waitForceKeepCh: make(chan struct{}),
+			keepFunc:        keepFunc,
+		}
+	}
+
+	go nf.loop()
+
+	log.Info("new node buffer list", "proposed block interval", nf.wpBlocks,
+		"reserve multi difflayers", nf.rsevMdNum, "difflayers in multidifflayer", nf.dlInMd,
+		"limit", common.StorageSize(limit), "layers", layers, "persist id", nf.persistID, "base_size", size)
+	return nf, nil
+}
+
+// newNodeBufferListForRecovery initializes the node buffer list for recovering diff layers
+func newNodeBufferListForRecovery(db ethdb.Database, freezer *rawdb.ResettableFreezer, base *multiDifflayer,
```

Comment:
Semantically, it is recover bufferlist, not new bufferlist for recover.

Reply:
Fixed